pkg/lib/cockpit/location.ts

```diff
@@ -0,0 +1,192 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2024 Red Hat, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+import { url_root, calculate_application } from './_internal/location-utils';
+
+/* HACK: Mozilla will unescape 'window.location.hash' before returning
+* it, which is broken.
+*
+* https://bugzilla.mozilla.org/show_bug.cgi?id=135309
+*/
+export function get_window_location_hash() {
+    return (window.location.href.split('#')[1] || '');
+}
+
+type Options = { [name: string]: string | Array<string> };
+type Path = string | string[] | Location;
```

Comment:
That surely is a confusing type. so "string" would be the whole raw path, like /my/page#hash?option1=value` ? Or even the whole URL (starting with `http://my.server`)? And `Location` is the fully parsed one. But what is string[] then? that smells like the path components, i.e. `["my", "page"]`, but where are the hash and options then?

Reply:
This should apply to `cockpit.location.go()` and `cockpit.location.replace` which with some grepping shows it takes a `string`:

```
pkg/systemd/terminal.jsx:            cockpit.location.replace("");
```

```
cockpit.location.go(cockpit.location.path, { x: "1" });
```

Or

```
onClick={ev => { ev.preventDefault(); cockpit.location.go(account.name) }}>
```

Reply:
Yes `string[]` is the "hash options":

```
onClick={() => cockpit.location.go(["vg", vgroup.Name])}>
```